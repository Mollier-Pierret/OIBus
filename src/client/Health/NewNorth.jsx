import React, { useState } from 'react'
import { Button, Popover, PopoverBody } from 'reactstrap'
import { nanoid } from 'nanoid'
import { ConfigContext } from '../context/configContext.jsx'
import NewApplicationRow from '../North/NewApplicationRow.jsx'
import validationNorth from '../North/Form/North.validation'

const NewNorth = () => {
  const { newConfig, dispatchNewConfig, apiList } = React.useContext(ConfigContext)
  const [openPopover, setOpenPopover] = useState(false)
  const applications = newConfig?.north?.applications ?? []

  const togglePopover = () => setOpenPopover(!openPopover)

  const addApplication = ({ name, api }) => {
    if (!applications.find((application) => application.name === name)) {
      togglePopover(false)
      dispatchNewConfig({
        type: 'addRow',
        name: 'north.applications',
        value: {
          id: nanoid(),
          name,
          api,
          enabled: false,
        },
      })
    }
  }

  return (
    <>
      <Button
        id="add-north"
        className="inline-button autosize oi-north-button"
        size="sm"
        outline
      >
        + North
      </Button>
      <Popover
        className="pop-container"
        trigger="legacy"
        placement="left"
        target="add-north"
        isOpen={openPopover}
        toggle={togglePopover}
      >
        <PopoverBody>
          <NewApplicationRow
            apiList={apiList}
            addApplication={addApplication}
            applicationList={applications?.map((application) => application.name) ?? []}
            valid={validationNorth.application.isValidName}
          />
        </PopoverBody>
      </Popover>
    </>

  )
}

export default NewNorth
