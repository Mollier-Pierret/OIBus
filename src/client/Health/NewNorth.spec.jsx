/**
 * @jest-environment jsdom
 */
import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

// need BrowserRouter so Link component is not complaining
import { BrowserRouter } from 'react-router-dom'

import NewNorth from './NewNorth.jsx'

import newConfig from '../../../tests/testConfig'

// ReacFlow does not seem to be working with jest.
// so we have to mock this component
jest.mock('../../../node_modules/react-flow-renderer/dist/ReactFlow.js', () => () => ('ReactFlow'))

let container

beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

React.useContext = jest.fn().mockReturnValue({ newConfig })
describe('NewNorth', () => {
  test('display NewNorth page based on config', async () => {
    act(() => {
      ReactDOM.render(
        <BrowserRouter>
          <NewNorth />
        </BrowserRouter>,
        container,
      )
    })
    expect(container).toMatchSnapshot()
  })
})
