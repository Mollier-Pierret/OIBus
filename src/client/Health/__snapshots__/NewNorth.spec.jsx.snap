// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewNorth display NewNorth page based on config 1`] = `
<div>
  <button
    class="inline-button autosize oi-north-button btn btn-outline-secondary btn-sm"
    id="add-north"
  >
    + North
  </button>
</div>
`;
